data/transactions/verify/verifiedTxnCache.go

```diff
@@ -0,0 +1,292 @@
+// already-pinned transactions. If a transaction is not currently pinned, and it's can't be found in the cache, a errMissingPinnedEntry error would be generated.
+func (v *verifiedTransactionCache) UpdatePinned(pinnedTxns map[transactions.Txid]transactions.SignedTxn) (err error) {
+	v.bucketsLock.Lock()
+	defer v.bucketsLock.Unlock()
+	pinned := make(map[transactions.Txid]*GroupContext, len(pinnedTxns))
+	for txID := range pinnedTxns {
+		if groupEntry, has := v.pinned[txID]; has {
+			pinned[txID] = groupEntry
+			continue
+		}
+
+		// entry isn't in pinned; maybe we have it in one of the buckets ?
+		found := false
+		// we use the (base + W) % W trick here so we can go backward and wrap around the zero.
+		for offsetBucketIdx := v.base + len(v.buckets); offsetBucketIdx > v.base; offsetBucketIdx-- {
+			bucketIdx := offsetBucketIdx % len(v.buckets)
+			if groupEntry, has := v.buckets[bucketIdx][txID]; has {
+				pinned[txID] = groupEntry
+				found = true
+				break
+			}
+		}
+		if !found {
+			err = errMissingPinnedEntry
+		}
+
+	}
+	v.pinned = pinned
+	return err
+}
+
+// Pin sets a given transaction group as pinned items, after they have already been verified.
+func (v *verifiedTransactionCache) Pin(txgroup []transactions.SignedTxn) (err error) {
+	v.bucketsLock.Lock()
+	defer v.bucketsLock.Unlock()
+	transcationMissing := false
+	if len(v.pinned)+len(txgroup) > maxPinnedEntries {
+		// reaching this number likely means that we have an issue not removing entries from the pinned map.
+		// return an error ( which would get logged )
+		return errTooManyPinnedEntries
+	}
+	baseBucket := v.base
+	for _, txn := range txgroup {
+		txID := txn.ID()
+		if _, has := v.pinned[txID]; has {
+			// it's already pinned; keep going.
+			continue
+		}
+
+		// entry isn't in pinned; maybe we have it in one of the buckets ?
+		found := false
+		// we use the (base + W) % W trick here so we can go backward and wrap around the zero.
+		for offsetBucketIdx := baseBucket + len(v.buckets); offsetBucketIdx > baseBucket; offsetBucketIdx-- {
+			bucketIdx := offsetBucketIdx % len(v.buckets)
+			if ctx, has := v.buckets[bucketIdx][txID]; has {
+				// move it to the pinned items :
+				v.pinned[txID] = ctx
+				delete(v.buckets[bucketIdx], txID)
+				found = true
+				baseBucket = bucketIdx
+				break
+			}
+		}
+		if !found {
+			transcationMissing = true
+		}
+	}
+	if transcationMissing {
+		err = errMissingPinnedEntry
+	}
+	return
+}
+
+// add is the internal implementation of Add/AddPayset which adds a transaction group to the buffer.
+func (v *verifiedTransactionCache) add(txgroup []transactions.SignedTxn, groupCtx *GroupContext) {
+	if len(v.buckets[v.base])+len(txgroup) > entriesPerBucket {
+		// move to the next bucket while deleting the content of the next bucket.
+		v.base = (v.base + 1) % len(v.buckets)
+		v.buckets[v.base] = make(map[transactions.Txid]*GroupContext, entriesPerBucket)
```

Comment:
probably it is better pre-allocate to max(entriesPerBucket, len(txgroup))

Reply:
the number of transaction in a group is 16. the entriesPerBucket is in the order of several thousands. 
when allocating a new bucket, we want to have large buckets, and have each bucket contain all the transactions of a single txn group.